lib/reservation_maker.rb

```diff
@@ -0,0 +1,80 @@
+require 'pry'
+require 'date'
+
+class ReservationMaker
+
+  attr_reader :start_date, :end_date
+
+  @@reservations = []
+
+  def initialize (start_date, end_date)
+    @start_date = start_date
+    @end_date = end_date
+
+    #Makes sure dates entered are realistic.  A start date after the end date is not.
+    if @end_date < @start_date
+      raise ArgumentError.new ("The start date cannot be after the end date.")
+    end
+  end
+
+  # This makes it so the @@reservations data is cumulative and keeps track of all reservations ever made with the program.
+  def self.reservations
+    @@reservations
+  end
+
+  # Dan added this part so that the above would work with the testing setup.
+  def self.reset
+    @@reservations = []
+  end
+
+  # lists all rooms in the hotel
+  def rooms
+    rooms = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
+    return rooms
+  end
+
+  # creates a list of all Booked Rooms (or rooms that are unavailable to be booked by the current customer) during the time frame requested.
+  def booked_rooms
+    booked_rooms = []
+    unless @@reservations.nil?
+      @@reservations.each do |reservation|
+        if @start_date < reservation.start_date && @end_date <= reservation.start_date || @start_date >= reservation.end_date
```

Comment:
I haven't totally sat down with this but... to me, this code looks like if this conditional evaluates to `true`, then... the dates (start date vs. reservation start date, end date vs. reservation end date) are NOT conflicting...

But this `if` says below that if this evaluates to true, then the reservation's room gets shoveled into "booked"!

I tested this by adding a `!` to the entire conditional, like so:
```ruby
if !(@start_date < reservation.start_date && @end_date <= reservation.start_date || @start_date >= reservation.end_date)
```

Let me know what you think! (There's a chance I overlooked something)

Reply:
If you end up having this conditional and then running the tests, then your failing tests turn into errors... because the test "it will NOT create another reservation because the 21st reservation is for the SAME DATES as the first 20 reservations." raises a `StandardError` with the message `StandardError: There are no rooms available for that time period.`... which makes sense ;)